### PR TITLE
LPS-98330 Add GraphQL contributions

### PR DESCRIPTION
**Pull request by javierdearcos**
Thin infrastructure to ease the use of GraphQL contributions to external clients.

This feature is already accessible to them but it is not very intuitive, mostly because of the naming.

With this infrastructure the client should implement a Component implementing the GraphQLContributor interface. This class should return an object which uses GraphQL annotations: GraphQLTypeExtension to select the class to contribute to and, inside the class,  GraphQLFields to add the contributed fields.

An example would be:
```
@Component(immediate = true, service = GraphQLContributor.class)
public class StructureContentGraphQLContributor implements GraphQLContributor {

	@Override
	public String getContributionPath() {
		return "/headless-delivery-graphql/v1_0";
	}

	@Override
	public Object getMutationContribution() {
		return null;
	}

	@Override
	public QueryContributor getQueryContribution() {
		return new QueryContributor();
	}

	public static class QueryContributor {

		@GraphQLTypeExtension(StructuredContent.class)
		public class MyStructuredContentTypeExtension {

			public MyStructuredContentTypeExtension(
				StructuredContent structuredContent) {

				_structuredContent = structuredContent;
			}

			@GraphQLField
			public String externalContribution() {
				return "External contribution of " +
					_structuredContent.getTitle();
			}

			private final StructuredContent _structuredContent;
		}
	}
}
```

### CODE REVIEW

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/graphql/servlet/GraphQLContributor.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+package com.liferay.portal.vulcan.graphql.servlet;
+
+/**
+ * @author Javier de Arcos
+ */
+public interface GraphQLContributor {
+
+	public Object getMutationContribution();
+
```
Comment (javierdearcos) [inline modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/graphql/servlet/GraphQLContributor.java:23]: I doubt about creating default methods returning null for getMutationContribution and getQueryContribution so the client only needs to implement the one that he need, and not add another implementation returning null

Comment (nhpatt) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/graphql/servlet/GraphQLContributor.java:23]: I would do it, it'll be cleaner that way.

Comment (javierdearcos) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/graphql/servlet/GraphQLContributor.java:23]: Great. I'll change it right now

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/graphql/servlet/GraphQLServletExtender.java [hunk 2]
```diff
@@ -2006,6 +2018,54 @@ public DataFetcherExceptionHandlerResult onException(
...
+				_servletDataServiceRegistrationMapping.remove(
+					graphQLContributor)
+			).ifPresent(
+				ServiceRegistration::unregister
+			);
+
+			_bundleContext.ungetService(serviceReference);
+		}
+
+		private final Map<GraphQLContributor, ServiceRegistration<ServletData>>
```
Comment (nhpatt) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/graphql/servlet/GraphQLServletExtender.java:2064]: I think this should be _servletDataServiceRegistrations or _servletDataServiceRegistrationMap

**Comment by nhpatt** [comment]
ci:forward

**Comment by javierdearcos** [comment]
ci:forward

**Pull request closed without merging**